ibmcloud provider

dchung-us opened the pull request:
Adding a new provider, ibmcloud,
this is for IBM Cloud.

shyamsabhaya commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_virtual_machine.py:
```diff
@@ -0,0 +1,841 @@
+    SUBNETX4: ['10.101.50.1', '10.102.50.1', '10.103.50.1', '10.104.50.1', '10.105.50.1']
+}
+
+USER_DATA = "Content-Type: multipart/mixed; boundary=MIMEBOUNDARY\n\
```
What is this USER_DATA used for? Is it necessary?

dchung-us replied:
First, thank you for reviewing the changes.
yes, USER_DATA is used for windows virtual machines to set up firewalls and ports used to run benchmarks between vm's.

shyamsabhaya commented on perfkitbenchmarker/providers/ibmcloud/utils.py:
```diff
@@ -0,0 +1,445 @@
+    return self.gen_vpcmgr.CreatePrefix(self.flags['vpcid'], self.flags['zone'],
+                                        self.flags['cidr'], name=self.flags['name'])
+
+  def CreateSubnet(self):
```
Should this be in IBM network?

dchung-us replied:
yes, this is needed in ibm cloud network, this is created on a vpc where subnets are attached, address prefix must be added to vpc before subnets can be attached to vpc.

shyamsabhaya commented on perfkitbenchmarker/providers/ibmcloud/utils.py:
```diff
@@ -0,0 +1,445 @@
+    resp = self.gen_keymgr.Create(self.flags['pubkey'], 'rsa', **kwargs)
+    return resp.get('id') if resp else None
+
+  def CreateInstance(self):
```
Utils should not create any instance or resource creation. Just helper functions that are used in multiple places and CLI use. See gcp_utils, aws_utils, or azure_utils.

shyamsabhaya replied:
Recommend splitting VM information into ibm_vm

dchung-us replied:
ok, makes sense, I will.

shyamsabhaya commented:
Also please provide documentation for functions.

dchung-us commented:
> Also please provide documentation for functions.

ok, I will split util as mentioned above, add more docs and submit the changes again,
thank you.

shyamsabhaya commented on CHANGES.next.md:
```diff
@@ -1,5 +1,6 @@
 ### Breaking changes:
 
+-   Add ibmcloud as a new provider
```
Please move this to `New Features`. This should not be a breaking change.

dchung-us replied:
yes, I moved it, thanks.

dchung-us commented:
Hello, could you please advise what the next step is on this ?

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/flags.py:
```diff
@@ -0,0 +1,121 @@
+flags.DEFINE_string('ibmcloud_region', 'us-south',
+                    'Name of IBM Cloud Location or region.')
+
+flags.DEFINE_string('ibmcloud_zones', 'us-south-1',
```
Can you use the regular FLAGS.zones ?  It would be nice to not have to not have to specify zones in a special way for a provider.

dchung-us replied:
yes I agree, updated.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/flags.py:
```diff
@@ -0,0 +1,121 @@
+flags.DEFINE_integer('ibmcloud_network_mtu', 9000,
+                     'MTU size on network interfaces.')
+
+flags.DEFINE_string('ibmcloud_image_name', None,
```
There's a FLAGS.image in pkb.py, can that be used here?

dchung-us replied:
yes.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/flags.py:
```diff
@@ -0,0 +1,121 @@
+flags.DEFINE_integer('ibmcloud_vdisks_extra', 0,
+                     'extra disks to create')
+
+flags.DEFINE_string('ibmcloud_preprovisioned_data', None,
```
The other ones are called "(cloudname)_preprovisioned_data_bucket"  can you change to that format?

dchung-us replied:
changed the name.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+class IbmCloudDisk(disk.BaseDisk):
+  """Object representing a IBM Cloud Disk"""
+
+  def __init__(self, disk_spec, name, zone):
```
For new code can you pytype it?  In this case

def __init__(self, disk_spec: disk.DiskSpec, name: str, zone:str)

I realize it looks like a pain to do but it really helps when looking at it later

dchung-us replied:
agree, thanks for the tip.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+  def _Create(self):
+    """Creates an external block volume."""
+    volcmd = ibm.IbmAPICommand(self)
+    volcmd.flags['name'] = self.name
```
Nit: I find it easier to read
  volcmd.flags.update(
{'name': self.name,
'zone' :self.zone,
....
'capacity' : self.data_disk_size,
}

dchung-us replied:
yes it is easier to read, updated.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+      vm: instance of the vm to which the disk will be attached.
+    """
+    if self.vol_id is None:
+      raise errors.Error('Cannot attach remote volume %s' % self.name)
```
Can you use f-strings here?
  f'Cannot attach remote volume {self.name}'

dchung-us replied:
changed to f string.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+    # wait till volume is PROVISIONED before attach
+    resp = json.loads(volcmd.ShowVolume())
+    status = resp['status']
+    endtime = time.time() + 300
```
Can you make this a constant?  In this case at the top of the file:
  _MAX_DISK_ATTACH_SECONDS = 300

endtime = time.time() + _MAX_DISK_ATTACH_SECONDS

dchung-us replied:
yes, added constants.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+    resp = volcmd.InstanceShowVolume()
+    logging.info('Show volume resp: %s', resp)
+    status = None
+    if resp:
```
I would just do the check inside the loop so you only have the code once, so remove this part and put the time.sleep(2) after you do the check (line 156)

dchung-us replied:
agree, just one call used.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/ibmcloud_disk.py:
```diff
@@ -0,0 +1,199 @@
+    while time.time() < endtime:
+      stdout, _ = self.attached_vm.RemoteCommand(cmd, should_log=True)
+      # parse for lines that contain "Disk /dev/vdd: 1000 GiB"
+      disks = re.findall(r'\Disk /.*/.*: \d+ GiB', stdout)
```
Can you do something like this?
 disks = re.findall(r'\Disk (\S+): (\d+) GiB', stdout)
 for disk_path, disk_size in disks:
    ...

dchung-us replied:
yes thanks for the tip, that is much better, updated.

shyamsabhaya commented:
Hi dchung,

I will be in charge of helping push this PR into Perfkitbenchmarker. Next steps needed to merge PR:

1. Please sync PR to head
2. Please add unit tests for the different resources (VM, network, etc - see other clouds as example) - this is to ensure that we do not break any IBM Cloud functionality when new code is added to PKB.
3. Please attach log files to the PR that show successful runs on the cluster_boot, netperf, and ping benchmarks.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+from perfkitbenchmarker import errors
+from perfkitbenchmarker.providers.ibmcloud import ibmcloud
+
+WINDOWS = 'windows'  # all windows
```
You should be able to use os_types.py for these except for centos (which is centos7 and centos8 in that file)

dchung-us replied:
I used some that exists in os_types, it doesn't have ubuntu either, so I left those.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+  """
+  try:
+    with open(config, 'r') as stream:
+      data = json.dumps(yaml.load(stream), sort_keys=True)
```
use `yaml.safe_load`

dchung-us replied:
yes, updated.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+    for i in range(len(ip_list)):
+      if ip_list[i] == ipv4_cidr_block:
+        return i
+  return -1
```
Might want to consider throwing an exception here if the caller expects to get a subnet.  If it is optional then note in the Returns that -1 is returned for missing.

dchung-us replied:
I just noted it returns -1 for missing.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+
+  Args:
+    data: output from route command on vm.
+    index: subnet index.
```
Seems a little odd to pass in an index to something when you can just pass in the subnet string, can you do that?

dchung-us replied:
for this one, I will have to leave as is for now, this is used for internal use only.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+  return route_cmds
+
+
+def GetMinMax(data):
```
pytype would really help here.  is it `(data:List[int]) -> Tuple[int,int]:` ?

dchung-us replied:
I agree, thanks for the tip, didn't really know about ptype but it really helps with readability.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+
+def GetMinMax(data):
+  """Finds min and max values in the given list of values"""
+  if not data:
```
data = [float(x) for x in data]
return (min(data), max(data))

dchung-us replied:
yes, actually this is not really needed so I just removed this function.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+
+def GetGen(account):
+  """Creates a ibmcloud access object """
+  gen = ibmcloud.IbmCloud(account=account[0], apikey=account[1], verbose=False, \
```
I'm a little leary of seeing references to tuple positions.  Can you do this?

```
@dataclasses.dataclass(frozen=True):
class Account:
   name:str
   apikey:str
```

then it is a lot easier to following:

```
def GetGen(account:Account) -> ?:
    get =   IbmCloud(account=account.name, apikey=account.apikey, ....)

dchung-us replied:
thanks for this tip here too, I don't like to use tuple indexes either ;)

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+      custom = True
+  else:
+    # if lookup failed, try read from env if any is set
+    data['version'] = os.environ.get('RIAS_IMAGE_OS_VERSION')
```
Do you need OS environment to be set?  That seems a bit brittle.  If you need them for your shell commands you can do this, assuming `ibm_cloud` is the executable

```
txt, _ = vm_util.IssueCommand(['ibm_cloud', 'foobar'],
env={'RIAS_IMAGE_OS_VERSION: 'foo'})
```

dchung-us replied:
We don't need to set OS env to run, it's only there to fill some properties if it is not found for some reason.

cwilkes commented on perfkitbenchmarker/providers/ibmcloud/util.py:
```diff
@@ -0,0 +1,212 @@
+    data['version'] = os.environ.get('RIAS_IMAGE_OS_VERSION')
+    data['vendor'] = os.environ.get('RIAS_IMAGE_OS_VENDOR')
+    data['name'] = os.environ.get('IMGNAME')
+    if data['name'] is not None and 'custom' in data['name']:
```
can just say `if 'custom' in data['name'] or ''`  that is assuming you want `data['name'] = None` to be a value.

dchung-us replied:
data['name'] can be None,
if 'custom' in data['name'] or ''
this fails so I am leaving as is.

cwilkes reviewed (commented):
initial pass

dchung-us commented:
@shyamsabhaya, @cwilkes 
Thank you for your advice and reviews, I will do what you are recommending and come back with updates.

dchung-us commented:
@shyamsabhaya, fyi, I have created some unittests, I will try to address the comments by @cwilkes next, may take some time before I submit all the changes.

dchung-us commented:
@shyamsabhaya, I think I have sync'd to head, I did git fetch upstream and git merge upstream/master after setting remote git with    git remote add upstream https://github.com/GoogleCloudPlatform/PerfKitBenchmarker.git.
Let me know if this is not the right way to sync.
I will upload the logs next from my test testruns.

@cwilkes, I addressed most of your comments and updated.

Also, I clicked on Resolve button for each comment, not sure if I am supposed to do that myself, let me know if not. Thanks!

I will aso be submitting a new PR with latest changes, changes per the review comments and some changes of my own.

dchung-us commented:
[ping.zip](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/files/5515095/ping.zip)
[cluster_boot.zip](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/files/5515096/cluster_boot.zip)
[netperf.zip](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/files/5515097/netperf.zip)

dchung-us commented:
@googlebot I fixed it.

dchung-us commented:
@shyamsabhaya , @cwilkes ,
I happened to use wrong email when I submitted my last PR and it failed CLA, I submitted a new PR with correct email but it is no longer showing up here, looks like it is stuck with the CLA error above.

I will try submitting a new PR, named ibmcloud2.

dchung-us commented:
@shyamsabhaya @cwilkes ,
please see the new pr, ibmcloud provider -2,
also, please let me know if there's a way to fix the CLA problem.

dlott commented:
Closing in favor of  https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/2583
